sdk/core/core/src/az_json_read.c

```diff
@@ -0,0 +1,394 @@
+#include <ctype.h>
+#include <math.h>
+
+inline bool az_json_is_white_space(az_option_byte const c) {
+  switch (c) {
+    case ' ':
+    case '\t':
+    case '\n':
+    case '\r':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_is_esc(az_option_byte const c) {
+  switch (c) {
+    case '\\':
+    case '"':
+    case '/':
+    case 'b':
+    case 'f':
+    case 'n':
+    case 'r':
+    case 't':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_is_e(az_option_byte const c) {
+  switch (c) {
+    case 'e':
+    case 'E':
+      return true;
+  }
+  return false;
+}
+
+inline az_result az_json_error_unexpected(az_option_byte const c) {
+  return c == AZ_OPTION_BYTE_NONE ? AZ_JSON_ERROR_UNEXPECTED_END : AZ_JSON_ERROR_UNEXPECTED_CHAR;
+}
+
+inline bool az_json_stack_is_empty(az_json_state const *const p) { return p->stack == 1; }
+
+inline az_json_stack_item az_json_stack_last(az_json_state const *const p) { return p->stack & 1; }
+
+inline az_result az_json_stack_push(az_json_state *const p_state, az_json_stack const stack) {
+  if (p_state->stack >> AZ_JSON_STACK_SIZE != 0) {
+    return AZ_JSON_ERROR_STACK_OVERFLOW;
+  }
+  p_state->stack = (p_state->stack << 1) | stack;
+  return AZ_OK;
+}
+
+inline az_result az_json_stack_pop(az_json_state *const p_state) {
+  if (p_state->stack <= 1) {
+    return AZ_JSON_ERROR_INVALID_STATE;
+  }
+  p_state->stack >>= 1;
+  return AZ_OK;
+}
+
+az_json_state az_json_state_create(az_const_span const buffer) {
+  return (az_json_state){
+      .reader = az_span_reader_create(buffer),
+      .stack = 1,
+  };
+}
+
+az_result az_json_read_expected_char(az_span_reader *const p_reader, uint8_t const expected) {
+  az_option_byte const c = az_span_reader_current(p_reader);
+  if (c != expected) {
+    return az_json_error_unexpected(c);
+  }
+  az_span_reader_next(p_reader);
+  return AZ_OK;
+}
+
+void az_json_read_white_space(az_span_reader *const p_reader) {
+  while (az_json_is_white_space(az_span_reader_current(p_reader))) {
+    az_span_reader_next(p_reader);
+  }
+}
+
+az_result az_json_read_keyword_rest(az_span_reader *const p_reader, az_const_span const keyword) {
+  az_span_reader_next(p_reader);
+  az_span_reader k = az_span_reader_create(keyword);
+  while (true) {
+    az_option_byte const ko = az_span_reader_current(&k);
+    if (ko == AZ_OPTION_BYTE_NONE) {
+      return AZ_OK;
+    }
+    az_option_byte const o = az_span_reader_current(p_reader);
+    if (o != ko) {
+      return az_json_error_unexpected(o);
+    }
+    az_span_reader_next(p_reader);
+    az_span_reader_next(&k);
+  }
+}
+
+// 18 decimal digits. 10^18 - 1.
+//                        0         1
+//                        012345678901234567
+#define AZ_DEC_NUMBER_MAX 999999999999999999ull
+
+typedef struct {
+  int sign;
+  uint64_t value;
+  bool remainder;
+  int16_t exp;
+} az_dec_number;
+
+double az_json_number_to_double(az_dec_number const *p) {
+  return p->value * pow(10, p->exp) * p->sign;
```

Comment:
this is probably vulnerable to overflow, needs tests

Reply:
yes, and it doesn't work as described in IEEE standard. It's a placeholder to use `atod` or another function.

Reply:
yeah, I think my previous comments got detached from the code at some point, so I'm being cranky about it again here :P